Add new API and update Ruby version dependency

davemcorwin opened the pull request:
This implementation assumes the proofer gem logic receives an 'applicant' of the format:
```
{
  first_name: { value: 'Dave', verified: true },
  last_name: { value: 'Corwin', verified: true },
  dob: { value: '01/01/2000', verified: false },
  ssn: { value: '111111111', verified: true },
}
```

I actually think a format that separates the data from the verification status might be better overall:
- it would result in less data transformation
- the PII would be separate from the verification information
- would allow more data to be kept with the verification information

davemcorwin committed:
Update minimum Ruby dependency and add new api

**WHY**
- Ease addition of new vendor implementations
- require forwardable is required in Ruby 2.3

davemcorwin committed:
Raise if required attributes missing

jmhooper commented on lib/proofer/v2/base.rb:
```diff
@@ -0,0 +1,78 @@
+        proofed_attrs = self.class.proofer.call(vendor_attrs, reasons)
+
+        # only update the attributes that which the vendor can access
+        proofed_applicant = update_applicant(applicant, proofed_attrs)
```
It looks like this will update the applicant will all of the attributes, and not just the ones the proofer can update? Or is that something we want to enforce in the subclass?

davemcorwin committed:
Update logic to most recent understanding
**WHY**
- use current applicant format to reduce change in IdP and simplify
- rename some methods/attributes to improve clarity
- simplified result and added places for all vendor information
- `add_error` and `add_message` return self for easy chaining
- explanation of `Result` data:
- errors: a set of strings that explains why proofing failed (determines success)
  - messages: a set of strings that does not impact success
  - context: any extra information we may include to help track proofing requests (ex. transaction id)
  - exception: a single unexpected exception that causes proofing to fail

davemcorwin committed:
Require v2/base and v2/result by default

davemcorwin committed:
Require vendor/vendor_base by default

davemcorwin committed:
Bunp version

davemcorwin committed:
Use a hash for errors

davemcorwin committed:
Bump version

davemcorwin committed:
Return plain objects for errors and messages, provide to_h

davemcorwin committed:
Evaluate proofer block in context of subclass instance

davemcorwin committed:
Update Gemfile.lock
